### PR TITLE
Support optional result indication when accepting an approval request

### PR DESCRIPTION
**Pull request by matthewwalsh0**
## Description

Update the `accept` method of the `ApprovalController` to return a promise which optionally only resolves once the approval request creator has indicated the success or failure of any post-approval logic.

This is achieved using callbacks optionally provided in the approval request promise.

This is necessary to support more complex approval flows in which the request acceptor, typically a client frontend, must know the success of the resulting logic, or must receive some resulting data. 

## Changes

- **ADDED**: `waitForResult` option to `accept` method to resolve new promise when success or failure is indicated. 
- **ADDED**: `expectsResult` option to `add` and `addAndShowApprovalRequest` method to change the resolved promise to return a result object including the `value` and new `resultCallbacks`.
- **ADDED**: `expectsResult` property to `ApprovalRequest` to store return type preference in state.
- **ADDED**: Associated types including `AddResult`, `AcceptResult`, `AcceptResultCallbacks`, and `AcceptOptions`.

## Checklist

- [x] I've updated the test suite for new or updated code as appropriate
- [x] I've updated documentation for new or updated code as appropriate (note: this will usually be JSDoc)
- [x] I've highlighted breaking changes using the "BREAKING" category above as appropriate

### CODE REVIEW

File: packages/approval-controller/src/ApprovalController.ts [hunk 1]
```diff
@@ -267,23 +323,48 @@ export class ApprovalController extends BaseControllerV2<
    * @param opts.requestData - Additional data associated with the request,
    * @param opts.requestState - Additional state associated with the request,
    * if any.
-   * @returns The approval promise.
+   * @returns The approval promise resolving to an AddResult object.
+   */
+  addAndShowApprovalRequest(
+    opts: AddApprovalOptions & { expectsResult: true },
+  ): Promise<AddResult>;
```
Comment (bergarces) [inline packages/approval-controller/src/ApprovalController.ts:330]: TIL this absolutely works:

https://www.typescriptlang.org/play?#code/C4TwDgpgBAggJnGYwCcD2A3AhgGwPJjACWaAdgM5QC8UA3gFBRNQQAekAxsOQEoTkBXHMAD8ALigAjNGhwQspADT0AvgG569UJFgI+g4dTqNmAMxkTywFEVIBzZes1wIHHFhTRTA0lxKkoLAQACjRCcgl4RGR0bHxCf0oAMjoWdldufSFgCWsBaBUASgkABXQAWyJyCAAeKKzhAD4NFzcPLx8-MkCQsO5IhCRUTFwCYjJyYqgytErqmp8Aa1I0AHdSZs0guGDaIoA6YAALCFJguPzqRqgOCdkIfZw0O3PcfP3zNELCjW3dtM4mX42Qkplw1SgB2Op1eOEuVGutwo90ez1h70+31+IVoAIyvGBwlyKEuUJOZwu0ARNzuclRL0pHxkWKAA

File: packages/approval-controller/src/ApprovalController.ts [hunk 2]
```diff
@@ -427,9 +529,42 @@ export class ApprovalController extends BaseControllerV2<
...
+   * is `false` or `undefined`.
    */
-  accept(id: string, value?: unknown): void {
-    this._deleteApprovalAndGetCallbacks(id).resolve(value);
+  accept(
+    id: string,
+    value?: unknown,
+    options?: AcceptOptions,
+  ): Promise<AcceptResult> {
+    const approval = this.get(id) as ApprovalRequest<any>;
```
Comment (bergarces) [inline packages/approval-controller/src/ApprovalController.ts:542]: I'd be explicit about why we are casting here.

If we move this line underneath `_deleteApprovalAndGetCallbacks` and add a comment saying that the previous method would have thrown if an approval doesn't exist for that id (although that's technically not true either, as here we are getting the `ApprovalRequest` from the state, but `_deleteApprovalAndGetCallbacks` gets the promise callbacks from a private field).

Also we can narrow it down to `as ApprovalRequest<ApprovalRequestData>`. In this case I'd prefer a `!`, but those are banned by the linter and that'd require an extra line so... `as` is fine.

Comment (bergarces) [reply packages/approval-controller/src/ApprovalController.ts:542]: A workaround that is probably not worth it in my opinion is to have `_deleteApprovalAndGetCallbacks` changed (both in name and shape) to return both the callbacks and the `ApprovalRequest` that is deleting, and throw if either of them is not present.

That would save you from having to cast anything.

Comment (bergarces) [reply packages/approval-controller/src/ApprovalController.ts:542]: It's actually not half bad once implemented, but I'll be happy with just a comment justifying the assertion.

```
accept(
    id: string,
    value?: unknown,
    options?: AcceptOptions,
  ): Promise<AcceptResult> {
const {
      approvalRequest,
      callbacks: { resolve: resolveRequestPromise },
    } = this._deleteApprovalAndGetRequestAndCallbacks(id);

...

private _deleteApprovalAndGetRequestAndCallbacks(id: string) {
    const approvalRequest = this.get(id);
    const callbacks = this._approvals.get(id);
    if (!approvalRequest || !callbacks) {
      throw new ApprovalRequestNotFoundError(id);
    }

    this._delete(id);
    return {
      callbacks,
      approvalRequest,
    };
  }
```

Comment (matthewwalsh0) [reply packages/approval-controller/src/ApprovalController.ts]: I thought about doing the same, but I think it's overkill for my use case.

I can't move the cast as it's deleted so the `get` call after would return undefined.

Added a comment though, and used the better type for the template.

**Review by bergarces — APPROVED** [review]

**Review by vinistevam — APPROVED** [review]

**Pull request merged**